### PR TITLE
Q030

### PR DESCRIPTION
**Pull request by lgdean**
addresses #248
addresses #439

I'd particularly like feedback on how to make this code readable enough to convince someone that it does the right things (and all the right things, particularly all the relevant test cases).  I wasn't sure how to structure it, to get all the contrasting pairs somewhat close to each other.

(For the prod code, I may also create a version that uses a bunch of `case`s, to see whether it ends up more readable/convincing.)

### CODE REVIEW

File: model/src/main/scala/hmda/model/institution/Institution.scala [hunk 1]
```diff
@@ -10,6 +10,7 @@ case class Institution(
     agency: Agency,
     institutionType: InstitutionType,
     hasParent: Boolean,
+    cra: Boolean = false, // TODO do we have this info when creating the institution? if so, then don't default here.
```
Comment (hkeeler) [inline model/src/main/scala/hmda/model/institution/Institution.scala]: I think the answer is _yes_.  CRA is a separate data source, but  we should be able to merge it with NIC whenever creating/updating Institution records.

**Comment by schbetsy** [comment]
The work you've done to make the `Q030` Spec readable is really slick. I like the nesting and the different implementations of `mustPass` and `mustFail`.

**Comment by hkeeler** [comment]
For any other reviewers, I found the output of the test to be really useful when trying to understand all the branches of logic here.

@lesgou, can you take a peek at this?  I can walk you through it if it doesn't make sense, or interpret what's going on under-the-hood in the code.

```
[info] Q030Spec:
[info] Q030
[info]   when action taken type is 7 or 8 (preapproval)
[info]   - must pass
[info]   when action taken is 1-6
[info]     when institution is NOT a CRA reporter
[info]       when all 4 geography fields are NA
[info]       - must pass
[info]       when county is present and small
[info]         when tract is NA
[info]         - must pass
[info]         when tract is present (and matches)
[info]         - must fail
[info]     when institution is a CRA reporter
[info]       when all 4 geography fields are NA
[info]       - must fail
[info]       when state or county is NA (no matter what else is true)
[info]       - must fail
[info]       when county is present and small
[info]         when tract is NA
[info]         - must pass
[info]         when tract is present (and matches)
[info]         - must pass
[info]     when property MSA/MD is present
[info]       when tract is NA
[info]         when MSA/MD, state and county match
[info]         - must pass
[info]         when MSA/MD, state and county do not match
[info]         - must fail
[info]         when county is large
[info]         - must fail
[info]       when tract is present
[info]         when MSA/MD, state, county, and tract match
[info]         - must pass
[info]         when MSA/MD, state, county, and tract do not match
[info]         - must fail
[info]     when property MSA/MD is NA
[info]       when tract is NA
[info]         when state and county match
[info]         - must pass
[info]         when state and county do not match
[info]         - must fail
[info]         when county is large
[info]         - must fail
[info]       when tract is present
[info]         when state, county, and tract match
[info]         - must pass
[info]         when state, county, and tract do not match
[info]         - must fail
[info]     when state is present but county is NA
[info]     - must fail when CRA is true
[info]     - must fail when CRA is false
[info]     when county is present but state is NA
[info]     - must fail when CRA is true
[info]     - must fail when CRA is false
```

**Comment by lesgou** [comment]
:+1: : wow, I think we've got it!

**Comment by hkeeler** [comment]
I think we're good here.  Merging...

**Pull request merged**